otherlibs/stdune/src/fpath.ml

```diff
@@ -19,30 +19,44 @@ type mkdir_p_result =
   | Already_exists
   | Created
 
-let rec mkdir_p ?perms t_s =
-  match mkdir ?perms t_s with
-  | Created -> Created
-  | Already_exists -> Already_exists
-  | Missing_parent_directory ->
-    if is_root t_s
-    then
-      Code_error.raise
-        "Impossible happened: [Fpath.mkdir] refused to create a directory at the root, \
-         allegedly because its parent was missing"
-        []
-    else (
-      let parent = Filename.dirname t_s in
-      match mkdir_p ?perms parent with
-      | Created | Already_exists ->
-        (* The [Already_exists] case might happen if some other process managed
-           to create the parent directory concurrently. *)
-        (match mkdir t_s ?perms with
-         | Created -> Created
-         | Already_exists -> Already_exists
-         | Missing_parent_directory ->
-           (* But we just successfully created the parent directory. So it was
-              likely deleted right now. Let's give up *)
-           Code_error.raise "failed to create parent directory" [ "t_s", Dyn.string t_s ]))
+let mkdir_p_result_to_dyn = function
+  | Already_exists -> Dyn.variant "Already_exists" []
+  | Created -> Dyn.variant "Created" []
+;;
+
+let mkdir_p ?perms full_t_s =
+  let rec loop t_s =
+    match mkdir ?perms t_s with
+    | Created -> Created
+    | Already_exists -> Already_exists
+    | Missing_parent_directory ->
+      if is_root t_s
+      then
+        Code_error.raise
+          "Impossible happened: [Fpath.mkdir] refused to create a directory at the root, \
+           allegedly because its parent was missing"
+          []
+      else (
+        let parent = Filename.dirname t_s in
+        match loop parent with
+        | (Created | Already_exists) as parent_mkdir_p_result ->
+          (* The [Already_exists] case might happen if some other process managed
+             to create the parent directory concurrently. *)
+          (match mkdir t_s ?perms with
+           | Created -> Created
+           | Already_exists -> Already_exists
+           | Missing_parent_directory ->
+             (* But we just successfully created the parent directory. So it was
+                likely deleted right now. Let's give up *)
+             Code_error.raise
+               "failed to create parent directory"
+               [ "tried to create directory", Dyn.string t_s
+               ; "missing parent", Dyn.string parent
+               ; "parent mkdir_p_result", mkdir_p_result_to_dyn parent_mkdir_p_result
+               ; "while creating direcotry", Dyn.string full_t_s
+               ]))
+  in
+  loop full_t_s
```

Comment:
it looks like in addition to the extra logging, you also refactored the function to use `loop`. can you remove this part to make the diff lighter if possible?